### ISSUE TITLE
Add `limit` parameters  to /api/v1/querying_exemplars

### ISSUE THREAD

**Issue by hanjm** [prometheus/prometheus#10107]
## Proposal
**Use case. Why is this important?**
The exemplars api may response very large cause grafana dashboard stuck. could it add a limit parameters to it?
https://prometheus.io/docs/prometheus/latest/querying/api/#querying-exemplars

*“Nice to have” is not a good use case. :)*

**Comment by roidelapluie** [comment]
Could you be more explicit? What would the limit do? Render the last examplars, a random subset, the first ones?

cc @cstyan

**Comment by cstyan** [comment]
We also don't have any kind of query result limit parameter for the regular query APIs.

**Comment by bboreham** [comment]
There is a hard-coded limit of 11,000 points for range queries.

**Comment by roidelapluie** [comment]
> There is a hard-coded limit of 11,000 points for range queries.

This is a limit of steps.

**Comment by dandv** [comment]
> We also don't have any kind of query result limit parameter for the regular query APIs.

Do I understand correctly that it's impossible to limit the number of raw samples returned by the [`query`](https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries) endpoint?

How can pagination of [raw samples](https://www.robustperception.io/extracting-raw-samples-from-prometheus) to be done (in chunks of equal size, as with `LIMIT` in InfluxQL)? Here's the [SO question](https://stackoverflow.com/questions/71849368/return-a-limited-number-of-prometheus-samples-between-start-and-end-timestamps).

**Comment by cstyan** [comment]
> > We also don't have any kind of query result limit parameter for the regular query APIs.
> 
> Do I understand correctly that it's impossible to limit the number of raw samples returned by the [`query`](https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries) endpoint?
> 
> How can pagination of [raw samples](https://www.robustperception.io/extracting-raw-samples-from-prometheus) to be done (in chunks of equal size, as with `LIMIT` in InfluxQL)? Here's the [SO question](https://stackoverflow.com/questions/71849368/return-a-limited-number-of-prometheus-samples-between-start-and-end-timestamps).

Correct. Prometheus' server side config contains fields to limit the number of samples that can be loaded into memory at once during the execution of a query, but there's no `limit` parameters in the `query` or similar endpoints.

**Comment by dandv** [comment]
Thanks for clarifying. Coming from Influx, this seems like a major oversight. Am I missing something?

**Comment by bboreham** [comment]
If reading a lot of raw samples in stages is what you want, then the [remote-read API](https://prometheus.io/blog/2019/10/10/remote-read-meets-streaming/) is more suitable.
Brian's blog post which you linked to is more about seeing some underlying data for debugging purposes.

As an Open Source project, Prometheus has those features that someone considered important enough to code up and contribute. I don't think "major oversight" applies.

**Comment by bboreham** [comment]
Recently I have hit this myself: if Prometheus returns thousands of exemplars then the client library will spend several seconds parsing them.

The client I'm using, Grafana, then down-samples the exemplars because it doesn't want to paint thousands of dots on the screen. So I start to think it would be better to down-sample on the server.

**Comment by yeya24** [comment]
+1 on this. Limit the number of returned exemplars would be nice.
There are several ways to do this. What I have in mind:
1. Introduce step to downsample exemplars?
2. Return K exemplars sorted by an attribute. Like most recent K exemplars, or top K exemplars with largest value.

**Comment by cstyan** [comment]
Without additional parameters or sampling I feel like a blanket hard limit on the # of exemplars returned by the query API would make them even less useful unless we also introduced pagination to that API endpoint. Not blocking someone from making improvements, but I still don't think a `limit` parameter is the right approach.

**Comment by beorn7** [comment]
Hello from the bug scrub!

Thanks for the discussion so far.
It would be good to come to a conclusion how to approach this so that somebody could pick this up and implement something.

It generally sounds like some random sampling or topk behavior is more desired than a plain 'limit'. We'll change the title accordingly

**Comment by krajorama** [comment]
Hello from the bug scrub, we'd like to add the explicit requirement for the returned limited number of exemplars to represent an even distribution from #16468 .
